src/imitation/algorithms/tabular_irl.py

```diff
@@ -171,294 +177,94 @@ def mce_irl(
-                "||grad||=%f, ||E[dr/dw]||=%f)"
+                "||grad||=%f)"
                 % (
                     t,
                     linf_delta,
-                    np.linalg.norm(rew_params),
-                    np.linalg.norm(grad),
-                    np.linalg.norm(pol_grad),
+                    weight_norm,
+                    grad_norm,
                 )
             )
 
-        # take a single optimiser step
-        opt_state = opt_update(t, grad, opt_state)
-        rew_params = opt_get_params(opt_state)
-        rmodel.set_params(rew_params)
         t += 1
 
-    return rew_params, visitations
-
+    return visitations
 
-# ############################### #
-# ####### REWARD MODELS ######### #
-# ############################### #
 
+class LinearRewardModel(nn.Module):
+    """Simplest possible linear reward model."""
 
-class RewardModel(abc.ABC):
-    """Abstract model for reward functions (linear, MLPs, nearest-neighbour, etc.)"""
-
-    @abc.abstractmethod
-    def out(self, inputs):
-        """Get rewards for a batch of observations.
-
-        Args:
-            inputs (np.ndarray): 2D matrix of observations, with first axis
-                most likely indexing over state & second indexing over elements
-                of observations themselves.
-
-        Returns:
-            np.ndarray of rewards (just a 1D vector with one element for each
-            supplied observation).
-        """
-
-    @abc.abstractmethod
-    def grads(self, inputs):
-        """Gradients of reward with respect to a batch of input observations.
+    def __init__(self, obs_dim: int):
+        """Construct a linear reward model, with no bias.
 
         Args:
-            inputs (np.ndarray): 2D matrix of observations, like .out().
-
-        Returns:
-            np.ndarray of gradients *with respect to each input separately*.
-            e.g if the model has a W-dimensional parameter vector, and there
-            are O observation passed in, then the return value will be an O*W
-            matrix of gradients.
+            obs_dim: size of input observations.
         """
+        super().__init__()
+        # no bias in this model for the sake of simplicity
+        self.lin = nn.Linear(obs_dim, 1, bias=False)
 
-    def out_grads(self, inputs):
-        """Combination method to do forward-prop AND back-prop. This is trivial for
-        linear models, but might provide some cost saving for deep ones.
+    def forward(self, obs: th.Tensor) -> th.Tensor:
+        return th.squeeze(self.lin(obs), 1)
 
-        Args:
-            inputs (np.ndarray): 2D matrix of observations, like .out().
 
-        Returns:
-            (np.ndarray, np.ndarray), where first array is equivalent to return
-            value of .out() and second array is equivalent to return value of
-            .grads().
-        """
-        return self.out(inputs), self.grads(inputs)
+class MLPRewardModel(nn.Module):
```

Comment:
This feels like it should be defined elsewhere and shared between algos, but it's not compatible with vanilla RewardNet, so I think we can keep it as is for this PR and then try to clean up in https://github.com/HumanCompatibleAI/imitation/pull/311